src/Compiler/CompileExpr.idr

```diff
@@ -135,38 +135,73 @@ mkDropSubst i es rest (x :: xs)
 -- NOTE: Make sure that names mentioned here are listed in 'natHackNames' in
 -- Common.idr, so that they get compiled, as they won't be spotted by the
 -- usual calls to 'getRefs'.
+data Magic : Type where
+  MagicCCon : Namespace -> String -> (arity : Nat) -> -- checks
+              (FC -> forall vars. Vect arity (CExp vars) -> CExp vars) -> -- translation
+              Magic
+  MagicCRef : Namespace -> String -> (arity : Nat) -> -- checks
+              (FC -> FC -> forall vars. Vect arity (CExp vars) -> CExp vars) -> --translation
+              Magic
+
+magic : List Magic -> CExp vars -> CExp vars
+magic ms (CLam fc x exp) = CLam fc x (magic ms exp)
+magic ms e = go ms e where
+
+  fire : Magic -> CExp vars -> Maybe (CExp vars)
+  fire (MagicCCon ns n arity f) (CCon fc (NS ns' (UN n')) _ es)
+    = do guard (n == n' && ns == ns')
+         map (f fc) (toVect arity es)
+  fire (MagicCRef ns n arity f) (CApp fc (CRef fc' (NS ns' (UN n'))) es)
+    = do guard (n == n' && ns == ns')
+         map (f fc fc') (toVect arity es)
+  fire _ _ = Nothing
+
+  go : List Magic -> CExp vars -> CExp vars
+  go [] e = e
+  go (m :: ms) e = case fire m e of
+    Nothing => go ms e
+    Just e' => e'
+
+natMinus : FC -> FC -> forall vars. Vect 2 (CExp vars) -> CExp vars
+natMinus fc fc' [m,n] = CApp fc (CRef fc' (UN "prim__sub_Integer")) [m, n]
+
 natHack : CExp vars -> CExp vars
-natHack (CCon fc (NS ["Types", "Prelude"] (UN "Z")) _ []) = CPrimVal fc (BI 0)
-natHack (CCon fc (NS ["Types", "Prelude"] (UN "S")) _ [k])
-    = CApp fc (CRef fc (UN "prim__add_Integer")) [CPrimVal fc (BI 1), k]
-natHack (CApp fc (CRef _ (NS ["Types", "Prelude"] (UN "natToInteger"))) [k]) = k
-natHack (CApp fc (CRef _ (NS ["Types", "Prelude"] (UN "integerToNat"))) [k]) = k
-natHack (CApp fc (CRef fc' (NS ["Types", "Prelude"] (UN "plus"))) args)
-    = CApp fc (CRef fc' (UN "prim__add_Integer")) args
-natHack (CApp fc (CRef fc' (NS ["Types", "Prelude"] (UN "mult"))) args)
-    = CApp fc (CRef fc' (UN "prim__mul_Integer")) args
-natHack (CApp fc (CRef fc' (NS ["Nat", "Data"] (UN "minus"))) args)
```

Comment:
HUM. This one is in `Data.Nat`!

Reply:
Oh my...

We really need to find a better way of doing this, for this and other reasons. I suppose that is why I chose the name `natHack` but still, the magic strings aren't good.

Pattern synonyms would help. They're high on my list of things to do, but term is starting soon...